# Bigmart Sales Data Set

Retail is another industry which extensively uses analytics to optimize business processes. Tasks like product placement, inventory management, customized offers, product bundling, etc. are being smartly handled using data science techniques. As the name suggests, this data comprises of transaction records of a sales store. This is a regression problem. The data has 8523 rows of 12 variables.

Problem: Predict the sales of a store.

Tutorial : https://www.analyticsvidhya.com/blog/2016/02/bigmart-sales-solution-top-20/

Get Data : https://datahack.analyticsvidhya.com/contest/practice-problem-big-mart-sales-iii/


The data scientists at BigMart have collected 2013 sales data for 1559 products across 10 stores in different cities. Also, certain attributes of each product and store have been defined. The aim is to build a predictive model and find out the sales of each product at a particular store.

Using this model, BigMart will try to understand the properties of products and stores which play a key role in increasing sales.

In [1]:
# imports
import pandas
import numpy
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [2]:
# Read Data
dataset=pandas.read_csv('Data/Train_BigMart.csv')

#### Summarize the Dataset

In [3]:
# 1. Dimensions of Dataset
print(dataset.shape)

(8523, 12)


In [4]:
# 2. Look at the data itself.
display(dataset.head(10))

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDX07       19.200          Regular         0.000000   
4           NCD19        8.930          Low Fat         0.000000   
5           FDP36       10.395          Regular         0.000000   
6           FDO10       13.650          Regular         0.012741   
7           FDP10          NaN          Low Fat         0.127470   
8           FDH17       16.200          Regular         0.016687   
9           FDU28       19.200          Regular         0.094450   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   
5           Baking Goods   51.4008            OUT018   
6            Snack Foods   57.6588            OUT013   
7            Snack Foods  107.7622            OUT027   
8           Frozen Foods   96.9726            OUT045   
9           Frozen Foods  187.8214            OUT017   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   
5                       2009      Medium               Tier 3   
6                       1987        High               Tier 3   
7                       1985      Medium               Tier 3   
8                       2002         NaN               Tier 2   
9                       2007         NaN               Tier 2   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052  
5  Supermarket Type2           556.6088  
6  Supermarket Type1           343.5528  
7  Supermarket Type3          4022.7636  
8  Supermarket Type1          1076.5986  
9  Supermarket Type1          4710.5350

#### Variable Description

Item_Identifier - Unique product ID

Item_Weight - Weight of product

Item_Fat_Content - Whether the product is low fat or not

Item_Visibility - The % of total display area of all products in a store allocated to the particular product

Item_Type - The category to which the product belongs

Item_MRP - Maximum Retail Price (list price) of the product

Outlet_Identifier - Unique store ID

Outlet_Establishment_Year - The year in which store was established

Outlet_Size - The size of the store in terms of ground area covered

Outlet_Location_Type - The type of city in which the store is located

Outlet_Type - Whether the outlet is just a grocery store or some sort of supermarket

Item_Outlet_Sales - Sales of the product in the particulat store. This is the outcome variable to be predicted.

 

In [5]:
# 3. Statistical Summary
display(dataset.describe()) # it will identify numeric data automatically

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [6]:
# 4. Check Missing Values
dataset.apply(lambda x: sum(x.isnull()),axis=0) # axis 0 means column wise. 1 means row wise.

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

We have 1463 missing values in Item_Weight and 2410 missing values in Outlet_Size

In [7]:
dataset[dataset['Item_Weight'].isnull()]
# 7th , 18th, 21st, 23rd , 29th (... so on) rows has Item_Weight NaN. 
# So we will go through data and check if there is any way to find out Item's weight or not.

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
36             FDV20          NaN          Regular         0.059512   
38             FDX10          NaN          Regular         0.123111   
39             FDB34          NaN          Low Fat         0.026481   
49             FDS02          NaN          Regular         0.255395   
59             FDI26          NaN          Low Fat         0.061082   
62             FDF09          NaN          Low Fat         0.012090   
63             FDY40          NaN          Regular         0.150286   
64             FDY45          NaN          Low Fat         0.026016   
77             FDN48          NaN          Low Fat         0.113720   
83             NCL18          NaN          Low Fat         0.293418   
84             FDR12          NaN          Regular         0.031382   
87             FDQ49          NaN          Regular         0.039058   
95             FDU04          NaN          Low Fat         0.009715   
98             NCP18          NaN          Low Fat         0.028460   
101            FDD10          NaN          Regular         0.045798   
108            DRE60          NaN          low fat         0.278974   
112            FDW39          NaN          Regular         0.064625   
113            FDX34          NaN          Low Fat         0.071637   
130            FDY25          NaN          Low Fat         0.033810   
132            NCR53          NaN          Low Fat         0.144338   
136            FDH35          NaN          Low Fat         0.059957   
153            DRK12          NaN          Low Fat         0.041683   
161            FDR07          NaN          Low Fat         0.077367   
168            NCB30          NaN          Low Fat         0.025579   
178            DRY23          NaN          Regular         0.191014   
...              ...          ...              ...              ...   
8349           FDN15          NaN          Low Fat         0.016653   
8351           NCE43          NaN               LF         0.102941   
8353           NCX41          NaN          Low Fat         0.031024   
8356           FDN45          NaN          Low Fat         0.117531   
8358           FDK27          NaN          Low Fat         0.015664   
8362           FDY37          NaN          Regular         0.026440   
8367           FDA32          NaN               LF         0.052691   
8372           FDA01          NaN          Regular         0.054115   
8373           FDX44          NaN          Low Fat         0.042758   
8375           NCM05          NaN          Low Fat         0.104784   
8383           NCQ54          NaN          Low Fat         0.012482   
8390           NCQ05          NaN          Low Fat         0.037829   
8404           DRH39          NaN          low fat         0.000000   
8405           FDB09          NaN          Low Fat         0.100493   
8422           FDD08          NaN          Low Fat         0.035183   
8435           FDT48          NaN          Low Fat         0.000000   
8442           FDX40          NaN          Low Fat         0.173324   
8444           FDO28          NaN          Low Fat         0.071948   
8449           FDG02          NaN          Low Fat         0.019717   
8454           NCH54          NaN          Low Fat         0.127234   
8458           FDX20          NaN          Low Fat         0.074518   
8463           FDG59          NaN          Low Fat         0.043025   
8469           FDQ45          NaN          Regular         0.019114   
8472           NCS41          NaN          Low Fat         0.053185   
8480           FDQ58         

In [8]:
dataset
# looking at 7th , 18th, 21st, 23rd , 29th (and so on) items, 
# it seems that we can find out the Item_Weight from Item_Identifier

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
5              FDP36       10.395          Regular         0.000000   
6              FDO10       13.650          Regular         0.012741   
7              FDP10          NaN          Low Fat         0.127470   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
10             FDY07       11.800          Low Fat         0.000000   
11             FDA03       18.500          Regular         0.045464   
12             FDX32       15.100          Regular         0.100014   
13             FDS46       17.600          Regular         0.047257   
14             FDF32       16.350          Low Fat         0.068024   
15             FDP49        9.000          Regular         0.069089   
16             NCB42       11.800          Low Fat         0.008596   
17             FDP49        9.000          Regular         0.069196   
18             DRI11          NaN          Low Fat         0.034238   
19             FDU02       13.350          Low Fat         0.102492   
20             FDN22       18.850          Regular         0.138190   
21             FDW12          NaN          Regular         0.035400   
22             NCB30       14.600          Low Fat         0.025698   
23             FDC37          NaN          Low Fat         0.057557   
24             FDR28       13.850          Regular         0.025896   
25             NCD06       13.000          Low Fat         0.099887   
26             FDV10        7.645          Regular         0.066693   
27             DRJ59       11.650          low fat         0.019356   
28             FDE51        5.925          Regular         0.161467   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8493           FDP21        7.420          Regular         0.025886   
8494           NCI54       15.200          Low Fat         0.000000   
8495           FDE22        9.695          Low Fat         0.029567   
8496           FDJ57        7.420          Regular         0.021696   
8497           FDT08       13.650          Low Fat         0.049209   
8498           NCP54       15.350          Low Fat         0.035293   
8499           NCK53       11.600          Low Fat         0.037574   
8500           NCQ42       20.350          Low Fat         0.000000   
8501           FDW21        5.340          Regular         0.005998   
8502           NCH43        8.420          Low Fat         0.070712   
8503           FDQ44       20.500          Low Fat         0.036133   
8504           NCN18          NaN          Low Fat         0.124111   
8505           FDB46       10.500          Regular         0.094146   
8506           DRF37       17.250          Low Fat         0.084676   
8507           FDN28        5.880          Regular         0.030242   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8510           FDN58       13.800          Regular         0.056862   
8511           FDF05       17.500          Low Fat         0.026980   
8512           FDR26       20.700          Low Fat         0.042801   
8513           FDH31       12.000          Regular         0.020407   
8514           FDA01       15.000          Regular         0.054489   
8515           FDH24       20.700          Low Fat         0.021518   
8516           NCJ19       18.600          Low Fat         0.118661   
8517           FDF53       20

In [9]:
# so lets cross check it. Lets check weather perticular Item_Identifier has same Item_Weight through out
dataset[dataset['Item_Identifier']=='FDW12']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
21             FDW12          NaN          Regular         0.035400   
1788           FDW12        8.315          Regular         0.059541   
2973           FDW12        8.315              reg         0.035627   
5319           FDW12        8.315          Regular         0.035543   
6285           FDW12        8.315          Regular         0.035572   
7004           FDW12        8.315          Regular         0.035644   
7259           FDW12        8.315          Regular         0.035773   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
21    Baking Goods  144.5444            OUT027                       1985   
1788  Baking Goods  143.6444            OUT010                       1998   
2973  Baking Goods  146.6444            OUT049                       1999   
5319  Baking Goods  143.3444            OUT013                       1987   
6285  Baking Goods  144.4444            OUT046                       1997   
7004  Baking Goods  147.1444            OUT045                       2002   
7259  Baking Goods  144.3444            OUT017                       2007   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
21        Medium               Tier 3  Supermarket Type3          4064.0432  
1788         NaN               Tier 3      Grocery Store           145.1444  
2973      Medium               Tier 1  Supermarket Type1           870.8664  
5319        High               Tier 3  Supermarket Type1          1596.5884  
6285       Small               Tier 1  Supermarket Type1          2902.8880  
7004         NaN               Tier 2  Supermarket Type1           870.8664  
7259         NaN               Tier 2  Supermarket Type1          3483.4656

In [10]:
# cross check
dataset[dataset['Item_Identifier']=='FDC37']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
23             FDC37          NaN          Low Fat         0.057557   
4652           FDC37         15.5          Low Fat         0.032846   
6212           FDC37         15.5          Low Fat         0.032924   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
23    Baking Goods  107.6938            OUT019                       1985   
4652  Baking Goods  106.7938            OUT013                       1987   
6212  Baking Goods  107.2938            OUT049                       1999   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
23         Small               Tier 1      Grocery Store           214.3876  
4652        High               Tier 3  Supermarket Type1          3215.8140  
6212      Medium               Tier 1  Supermarket Type1          2143.8760

In [11]:
# cross check
dataset[dataset['Item_Identifier']=='FDC14']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility Item_Type  \
29             FDC14          NaN          Regular         0.072222    Canned   
3609           FDC14         14.5          Regular         0.041482    Canned   
4297           FDC14         14.5          Regular         0.069042    Canned   
5760           FDC14         14.5          Regular         0.000000    Canned   
6257           FDC14          NaN          Regular         0.041049    Canned   
6346           FDC14         14.5          Regular         0.041215    Canned   
6478           FDC14         14.5          Regular         0.041249    Canned   
7031           FDC14         14.5          Regular         0.041313    Canned   

      Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
29     43.6454            OUT019                       1985       Small   
3609   40.0454            OUT017                       2007         NaN   
4297   41.4454            OUT010                       1998         NaN   
5760   41.0454            OUT018                       2009      Medium   
6257   41.2454            OUT027                       1985      Medium   
6346   42.0454            OUT013                       1987        High   
6478   40.5454            OUT046                       1997       Small   
7031   42.0454            OUT049                       1999      Medium   

     Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
29                 Tier 1      Grocery Store           125.8362  
3609               Tier 2  Supermarket Type1           209.7270  
4297               Tier 3      Grocery Store            41.9454  
5760               Tier 3  Supermarket Type2           545.2902  
6257               Tier 3  Supermarket Type3          1342.2528  
6346               Tier 3  Supermarket Type1           629.1810  
6478               Tier 1  Supermarket Type1          1216.4166  
7031               Tier 1  Supermarket Type1           377.5086

In [12]:
item_avg_weight = dataset.pivot_table(values='Item_Weight', index='Item_Identifier')

In [13]:
# Lets cross check weather FDC14 has value 14.5 or not
item_avg_weight.T['FDC14']

Item_Weight    14.5
Name: FDC14, dtype: float64

In [14]:
item_avg_weight.head()
# Lets join this item_weights to the actual dataset item_weights by Item_Identifier

Item_Weight
Item_Identifier             
DRA12                 11.600
DRA24                 19.350
DRA59                  8.270
DRB01                  7.390
DRB13                  6.115

In [15]:
dataset=pandas.merge(dataset,item_avg_weight,on='Item_Identifier',how='left')

In [16]:
dataset = dataset.drop(columns="Item_Weight_x")
dataset = dataset.rename(columns={"Item_Weight_y": "Item_Weight"})

In [17]:
dataset.head()

Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
0           FDA15          Low Fat         0.016047                  Dairy   
1           DRC01          Regular         0.019278            Soft Drinks   
2           FDN15          Low Fat         0.016760                   Meat   
3           FDX07          Regular         0.000000  Fruits and Vegetables   
4           NCD19          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998         NaN   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  Item_Weight  
0               Tier 1  Supermarket Type1          3735.1380         9.30  
1               Tier 3  Supermarket Type2           443.4228         5.92  
2               Tier 1  Supermarket Type1          2097.2700        17.50  
3               Tier 3      Grocery Store           732.3800        19.20  
4               Tier 3  Supermarket Type1           994.7052         8.93

In [18]:
#Check Missing Values
dataset.apply(lambda x: sum(x.isnull()),axis=0) # axis 0 means column wise. 1 means row wise.

Item_Identifier                 0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
Item_Weight                     4
dtype: int64

In [19]:
dataset[dataset['Item_Weight'].isnull()]
# still we have some null values

Item_Identifier Item_Fat_Content  Item_Visibility     Item_Type  \
927            FDN52          Regular         0.130933  Frozen Foods   
1922           FDK57          Low Fat         0.079904   Snack Foods   
4187           FDE52          Regular         0.029742         Dairy   
5022           FDQ60          Regular         0.191501  Baking Goods   

      Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
927    86.9198            OUT027                       1985      Medium   
1922  120.0440            OUT027                       1985      Medium   
4187   88.9514            OUT027                       1985      Medium   
5022  121.2098            OUT019                       1985       Small   

     Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  Item_Weight  
927                Tier 3  Supermarket Type3          1569.9564          NaN  
1922               Tier 3  Supermarket Type3          4434.2280          NaN  
4187               Tier 3  Supermarket Type3          3453.5046          NaN  
5022               Tier 1      Grocery Store           120.5098          NaN

In [20]:
dataset[dataset['Item_Identifier']=='FDN52']

Item_Identifier Item_Fat_Content  Item_Visibility     Item_Type  Item_MRP  \
927           FDN52          Regular         0.130933  Frozen Foods   86.9198   

    Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
927            OUT027                       1985      Medium   

    Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  Item_Weight  
927               Tier 3  Supermarket Type3          1569.9564          NaN

In [21]:
# filling NA by mean
dataset['Item_Weight'].fillna(dataset['Item_Weight'].mean(), inplace=True)

In [22]:
#Check Missing Values
dataset.apply(lambda x: sum(x.isnull()),axis=0) # axis 0 means column wise. 1 means row wise.

Item_Identifier                 0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
Item_Weight                     0
dtype: int64

In [23]:
# there are 2410 rows which has null values for outlet_size
dataset[dataset['Outlet_Size'].isnull()].head()


Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
3            FDX07          Regular         0.000000  Fruits and Vegetables   
8            FDH17          Regular         0.016687           Frozen Foods   
9            FDU28          Regular         0.094450           Frozen Foods   
25           NCD06          Low Fat         0.099887              Household   
28           FDE51          Regular         0.161467                  Dairy   

    Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
3   182.0950            OUT010                       1998         NaN   
8    96.9726            OUT045                       2002         NaN   
9   187.8214            OUT017                       2007         NaN   
25   45.9060            OUT017                       2007         NaN   
28   45.5086            OUT010                       1998         NaN   

   Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  Item_Weight  
3                Tier 3      Grocery Store           732.3800       19.200  
8                Tier 2  Supermarket Type1          1076.5986       16.200  
9                Tier 2  Supermarket Type1          4710.5350       19.200  
25               Tier 2  Supermarket Type1           838.9080       13.000  
28               Tier 3      Grocery Store           178.4344        5.925

In [24]:
# we can find out outlet_size from outlet_Identifier

In [25]:
dataset[dataset['Outlet_Identifier']=='OUT017'].head()

Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
9            FDU28          Regular         0.094450           Frozen Foods   
25           NCD06          Low Fat         0.099887              Household   
53           FDA43          Low Fat         0.065042  Fruits and Vegetables   
54           NCP18          Low Fat         0.028760              Household   
73           FDG02          Low Fat         0.011325                 Canned   

    Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
9   187.8214            OUT017                       2007         NaN   
25   45.9060            OUT017                       2007         NaN   
53  196.3794            OUT017                       2007         NaN   
54  151.4708            OUT017                       2007         NaN   
73  189.6188            OUT017                       2007         NaN   

   Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  Item_Weight  
9                Tier 2  Supermarket Type1          4710.5350       19.200  
25               Tier 2  Supermarket Type1           838.9080       13.000  
53               Tier 2  Supermarket Type1          3121.2704       10.895  
54               Tier 2  Supermarket Type1          4815.0656       12.150  
73               Tier 2  Supermarket Type1          2285.0256        7.855

In [26]:
#dataset.pivot_table(values='Outlet_Size', index='Outlet_Identifier')
from scipy.stats import mode
dataset.pivot_table(values='Outlet_Size', columns='Outlet_Identifier',aggfunc=(lambda x:mode(x).mode[0]) ).T

/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


Outlet_Size
Outlet_Identifier            
OUT010                      0
OUT013                   High
OUT017                      0
OUT018                 Medium
OUT019                  Small
OUT027                 Medium
OUT035                  Small
OUT045                      0
OUT046                  Small
OUT049                 Medium

In [27]:
(dataset['Outlet_Identifier']).unique()

array(['OUT049', 'OUT018', 'OUT010', 'OUT013', 'OUT027', 'OUT045',
       'OUT017', 'OUT046', 'OUT035', 'OUT019'], dtype=object)

In [28]:
dataset[dataset['Outlet_Size'].isnull()]['Outlet_Identifier'].unique()

array(['OUT010', 'OUT045', 'OUT017'], dtype=object)

In [29]:
#Lets impute this later

#dataset = dataset.drop(columns=["Outlet_Size"])

In [30]:
dataset = dataset.drop(columns=["Item_Identifier",'Outlet_Identifier'])

In [31]:
dataset[['Item_Type','Item_Outlet_Sales']].groupby('Item_Type').agg(['mean']) # mean overall 2100
dataset=dataset.drop(columns=['Item_Type'])

In [32]:
dataset[['Outlet_Type','Item_Outlet_Sales']].groupby('Outlet_Type').agg(['median','min','max']) # mean overall 2100
# It seems that there is relation between Outlet_type and Item_Outlet_Sales
# Glocery store has less Item_Outlet_Sales
# Supermarket type 1 and 2 has almost equal Item_Outlet_Sales
# Supermarket 3 has high Item_Outlet_Sales

Item_Outlet_Sales                      
                             median       min         max
Outlet_Type                                              
Grocery Store              256.9988   33.2900   1775.6886
Supermarket Type1         1990.7420   73.2380  10256.6490
Supermarket Type2         1655.1788   69.2432   6768.5228
Supermarket Type3         3364.9532  241.6854  13086.9648

In [33]:
dataset['Outlet_Type']=dataset['Outlet_Type'].str.replace('Grocery Store','2')
dataset['Outlet_Type']=dataset['Outlet_Type'].str.replace('Supermarket Type1','19')
dataset['Outlet_Type']=dataset['Outlet_Type'].str.replace('Supermarket Type2','16')
dataset['Outlet_Type']=dataset['Outlet_Type'].str.replace('Supermarket Type3','34')
dataset['Outlet_Type']=pandas.to_numeric(dataset['Outlet_Type'], errors='coerce')

In [34]:
dataset[['Outlet_Size','Item_Outlet_Sales']].groupby('Outlet_Size').agg(['median','min','max']) # mean overall 2100

Item_Outlet_Sales                     
                       median      min         max
Outlet_Size                                       
High                2050.6640  73.2380  10256.6490
Medium              2251.0698  69.2432  13086.9648
Small               1544.6560  33.9558   9779.9362

In [35]:
dataset['Outlet_Size']=dataset['Outlet_Size'].str.replace('High','20')
dataset['Outlet_Size']=dataset['Outlet_Size'].str.replace('Medium','22')
dataset['Outlet_Size']=dataset['Outlet_Size'].str.replace('Small','15')
dataset['Outlet_Size']=pandas.to_numeric(dataset['Outlet_Size'], errors='coerce')
dataset['Outlet_Size'].fillna(dataset['Outlet_Size'].mean(), inplace=True)

In [36]:
dataset[['Item_Fat_Content','Item_Outlet_Sales']].groupby('Item_Fat_Content').agg(['median','min','max','count']) # mean overall 2100

Item_Outlet_Sales                           
                            median      min         max count
Item_Fat_Content                                             
LF                       1797.6600  40.6138   7503.5660   316
Low Fat                  1765.7016  33.2900  13086.9648  5089
Regular                  1845.5976  33.9558  12117.5600  2889
low fat                  1614.2321  83.8908   9562.2196   112
reg                      1492.7236  37.9506   7833.8028   117

In [37]:
dataset['Item_Fat_Content']=dataset['Item_Fat_Content'].str.replace('LF','17')
dataset['Item_Fat_Content']=dataset['Item_Fat_Content'].str.replace('Low Fat','17')
dataset['Item_Fat_Content']=dataset['Item_Fat_Content'].str.replace('low fat','16')
dataset['Item_Fat_Content']=dataset['Item_Fat_Content'].str.replace('Regular','18')
dataset['Item_Fat_Content']=dataset['Item_Fat_Content'].str.replace('reg','14')
dataset['Item_Fat_Content']=pandas.to_numeric(dataset['Item_Fat_Content'], errors='coerce')

In [38]:
dataset[['Outlet_Location_Type','Item_Outlet_Sales']].groupby('Outlet_Location_Type').agg(['median','min','max','count']) # mean overall 2100

Item_Outlet_Sales                           
                                median      min         max count
Outlet_Location_Type                                             
Tier 1                       1487.3972  33.9558   9779.9362  2388
Tier 2                       2004.0580  99.8700   9664.7528  2785
Tier 3                       1812.3076  33.2900  13086.9648  3350

In [39]:
dataset['Outlet_Location_Type']=dataset['Outlet_Location_Type'].str.replace('Tier 1','1487')
dataset['Outlet_Location_Type']=dataset['Outlet_Location_Type'].str.replace('Tier 2','2004')
dataset['Outlet_Location_Type']=dataset['Outlet_Location_Type'].str.replace('Tier 3','1812')

dataset['Outlet_Location_Type']=pandas.to_numeric(dataset['Outlet_Location_Type'], errors='coerce')

In [40]:
dataset=pandas.get_dummies(dataset, 
                                 prefix=None, prefix_sep='_', 
                                 dummy_na=False, columns=None, sparse=False, drop_first=False, dtype=None)

In [41]:
dataset.head().T

0            1           2            3  \
Item_Fat_Content             17.000000    18.000000    17.00000    18.000000   
Item_Visibility               0.016047     0.019278     0.01676     0.000000   
Item_MRP                    249.809200    48.269200   141.61800   182.095000   
Outlet_Establishment_Year  1999.000000  2009.000000  1999.00000  1998.000000   
Outlet_Size                  22.000000    22.000000    22.00000    18.960576   
Outlet_Location_Type       1487.000000  1812.000000  1487.00000  1812.000000   
Outlet_Type                  19.000000    16.000000    19.00000     2.000000   
Item_Outlet_Sales          3735.138000   443.422800  2097.27000   732.380000   
Item_Weight                   9.300000     5.920000    17.50000    19.200000   

                                   4  
Item_Fat_Content             17.0000  
Item_Visibility               0.0000  
Item_MRP                     53.8614  
Outlet_Establishment_Year  1987.0000  
Outlet_Size                  20.0000  
Outlet_Location_Type       1812.0000  
Outlet_Type                  19.0000  
Item_Outlet_Sales           994.7052  
Item_Weight                   8.9300

# Build Models

In [42]:
model_DF=dataset

In [43]:
from sklearn import model_selection # for splitting into train and test
import sklearn
# Split-out validation dataset
X = model_DF.drop(columns='Item_Outlet_Sales')
Y = model_DF['Item_Outlet_Sales']

validation_size = 0.20
seed = 100
X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [44]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [45]:
model_LR = LinearRegression()
model_LR.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [46]:
trainResult_LR=model_LR.predict(X_train)
testResult_LR=model_LR.predict(X_test)

In [47]:
def adj_r2_score(model,actual,predicted):
        from sklearn import metrics
        adj = 1 - float(len(actual)-1)/(len(actual)-len(model.coef_)-1)*(1 - metrics.r2_score(actual,predicted))
        return adj
    
    

########## TRAIN DATA RESULT ##########

print('---------- TRAIN DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_train, trainResult_LR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_train, trainResult_LR))


print('Adj R2 score: %.4f' % adj_r2_score(model_LR,Y_train, trainResult_LR))

########## TEST DATA RESULT ##########

print('---------- TEST DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_test, testResult_LR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_test, testResult_LR))

print('Adj R2 score: %.4f' % adj_r2_score(model_LR,Y_test, testResult_LR))


---------- TRAIN DATA RESULT ----------
Mean squared error: 1262157
Variance score: 0.5650
Adj R2 score: 0.5645
---------- TEST DATA RESULT ----------
Mean squared error: 1332354
Variance score: 0.5484
Adj R2 score: 0.5463


In [48]:
model_RidgeLR = Ridge(alpha=0.05)
model_RidgeLR.fit(X_train,Y_train)

trainResult_RidgeLR=model_RidgeLR.predict(X_train)
testResult_RidgeLR=model_RidgeLR.predict(X_test)

########## TRAIN DATA RESULT ##########

print('---------- TRAIN DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_train, trainResult_RidgeLR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_train, trainResult_RidgeLR))


print('Adj R2 score: %.4f' % adj_r2_score(model_RidgeLR,Y_train, trainResult_RidgeLR))


########## TEST DATA RESULT ##########
print('---------- TEST DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_test, testResult_RidgeLR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_test, testResult_RidgeLR))

print('Adj R2 score: %.4f' % adj_r2_score(model_RidgeLR,Y_test, testResult_RidgeLR))

---------- TRAIN DATA RESULT ----------
Mean squared error: 1262157
Variance score: 0.5650
Adj R2 score: 0.5645
---------- TEST DATA RESULT ----------
Mean squared error: 1332356
Variance score: 0.5484
Adj R2 score: 0.5463


In [49]:
model_LassoLR = Lasso(alpha=0.05)
model_LassoLR.fit(X_train,Y_train)

trainResult_LassoLR=model_LassoLR.predict(X_train)
testResult_LassoLR=model_LassoLR.predict(X_test)


########## TRAIN DATA RESULT ##########

print('---------- TRAIN DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_train, trainResult_LassoLR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_train, trainResult_LassoLR))


print('Adj R2 score: %.4f' % adj_r2_score(model_LassoLR,Y_train, trainResult_LassoLR))


########## TEST DATA RESULT ##########

print('---------- TEST DATA RESULT ----------')
# The mean squared error
print("Mean squared error: %.0f"% mean_squared_error(Y_test, testResult_LassoLR))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % r2_score(Y_test, testResult_LassoLR))

print('Adj R2 score: %.4f' % adj_r2_score(model_LassoLR,Y_test, testResult_LassoLR))

---------- TRAIN DATA RESULT ----------
Mean squared error: 1262158
Variance score: 0.5650
Adj R2 score: 0.5645
---------- TEST DATA RESULT ----------
Mean squared error: 1332389
Variance score: 0.5484
Adj R2 score: 0.5463


In [50]:
from scipy import stats

X_Columns=model_DF.drop(columns=['Item_Outlet_Sales']).columns # Name of all columns without target column

tempList=['Intercept']
tempList.extend(X_Columns)

params = numpy.append(model_LR.intercept_,model_LR.coef_)
predictions = model_LR.predict(X_train)

newX = pandas.DataFrame({"Constant":numpy.ones(len(X))}).join(pandas.DataFrame(X))
MSE = mean_squared_error(Y_test, testResult_LR)

var_b = MSE*(numpy.linalg.inv(numpy.dot(newX.T,newX)).diagonal())
sd_b = numpy.sqrt(var_b)
ts_b = params/ sd_b

p_values =[2*(1-stats.t.cdf(numpy.abs(i),(len(newX)-1))) for i in ts_b]

sd_b = numpy.round(sd_b,3)
ts_b = numpy.round(ts_b,3)
p_values = numpy.round(p_values,12)
params = numpy.round(params,4)

myDF3 = pandas.DataFrame()
myDF3["Cols"],myDF3["Coefficients"],myDF3["Standard Errors"],myDF3["t values"],myDF3["Probabilites"] = [tempList,params,sd_b,ts_b,p_values]
display(myDF3)

Cols  Coefficients  Standard Errors  t values  \
0                  Intercept   -17158.6109         3373.100    -5.087   
1           Item_Fat_Content       47.2266           19.928     2.370   
2            Item_Visibility     -304.5119          251.306    -1.212   
3                   Item_MRP       15.5627            0.201    77.473   
4  Outlet_Establishment_Year        7.3235            1.702     4.303   
5                Outlet_Size       -5.0041            4.933    -1.014   
6       Outlet_Location_Type       -0.0808            0.070    -1.149   
7                Outlet_Type      107.2892            1.880    57.081   
8                Item_Weight        0.5288            2.694     0.196   

   Probabilites  
0  3.717310e-07  
1  1.781646e-02  
2  2.256549e-01  
3  0.000000e+00  
4  1.701341e-05  
5  3.103838e-01  
6  2.504724e-01  
7  0.000000e+00  
8  8.443754e-01